File: src/client/net/net-client.ts
/**
 * The client half of the connection.
 *
 * Owns the socket, the prediction buffer and the snapshot history, and exposes
 * the result as plain state the rest of the client reads — same one-directional
 * contract as everything else here: the network writes into the world, and
 * nothing downstream writes back.
 *
 * The local player and everyone else are handled completely differently, and
 * that asymmetry is the whole design:
 *
 *   - **You** are predicted. Your input moves you on the frame you pressed it,
 *     and the server's correction is folded in when it arrives.
 *   - **Everyone else** is interpolated, rendered a fifth of a second in the
 *     past so their motion is smooth between the twenty snapshots a second that
 *     describe it.
 */

import { NET, TICK_DT } from '../../shared/constants.js';
import {
  NetMessage,
  decodeControl,
  decodeSnapshot,
  encodeControl,
  encodeInputs,
  peekType,
  type ByePayload,
  type ChatPayload,
  type RejectPayload,
  type WelcomePayload,
  type WireInput,
} from '../../shared/net/protocol.js';
import type { CollisionWorld } from '../../shared/collision/collision-types.js';
import type { InputCommand, PlayerId, PlayerState, SimEvent } from '../../shared/types.js';
import type { Loadout } from '../../shared/sim/loadout.js';
import { Predictor, type PredictionStats } from './prediction.js';
import { SnapshotBuffer } from './snapshot-buffer.js';

export type NetStatus = 'connecting' | 'playing' | 'rejected' | 'disconnected';

export interface NetClientOptions {
  url: string;
  name: string;
  loadout: Loadout;
  collision: CollisionWorld;
}

export class NetClient {
  status: NetStatus = 'connecting';
  /** Why the connection ended, when it did. */
  statusDetail = '';
  welcome: WelcomePayload | null = null;
  localId: PlayerId = 0;

  readonly snapshots = new SnapshotBuffer();
  readonly chat: ChatPayload[] = [];
  /** Events from the server, drained by the client each frame. */
  readonly events: SimEvent[] = [];

  private socket: WebSocket | null = null;
  private readonly predictor: Predictor;
  private seq = 1;
  /** Round-trip time in seconds, smoothed. */
  private rtt = 0;
  private pingTimer = 0;
  /** Local time elapsed since the newest snapshot, for interpolation. */
  private sinceSnapshot = 0;

  constructor(private readonly opts: NetClientOptions) {
    this.predictor = new Predictor(opts.collision);
    this.open();
  }

  private open(): void {
    const socket = new WebSocket(this.opts.url);
    socket.binaryType = 'arraybuffer';
    this.socket = socket;

    socket.onopen = () => {
      this.send(
        encodeControl(NetMessage.Hello, {
          protocolVersion: NET.protocolVersion,
          name: this.opts.name,
          loadout: this.opts.loadout,
        }),
      );
    };

    socket.onmessage = (ev) => this.receive(new Uint8Array(ev.data as ArrayBuffer));

    socket.onclose = () => {
      if (this.status !== 'rejected') {
        this.status = 'disconnected';
        this.statusDetail ||= '連線中斷';
      }
    };
    socket.onerror = () => {
      if (this.status === 'connecting') {
        this.status = 'rejected';
        this.statusDetail = '無法連線到伺服器';
      }
    };
  }

  // -------------------------------------------------------------------------

  private receive(bytes: Uint8Array): void {
    switch (peekType(bytes)) {
      case NetMessage.Welcome: {
        this.welcome = decodeControl<WelcomePayload>(bytes).payload;
        this.localId = this.welcome.yourId;
        this.status = 'playing';
        break;
      }
      case NetMessage.Reject: {
        this.status = 'rejected';
        this.statusDetail = decodeControl<RejectPayload>(bytes).payload.reason;
        break;
      }
      case NetMessage.Snapshot: {
        this.snapshots.push(decodeSnapshot(bytes));
        this.sinceSnapshot = 0;
        break;
      }
      case NetMessage.Events: {
        for (const e of decodeControl<SimEvent[]>(bytes).payload) this.events.push(e);
        break;
      }
      case NetMessage.Chat: {
        this.chat.push(decodeControl<ChatPayload>(bytes).payload);
        if (this.chat.length > 32) this.chat.shift();
        break;
      }
      case NetMessage.Bye: {
        const { id } = decodeControl<ByePayload>(bytes).payload;
        if (id === this.localId) {
          this.status = 'disconnected';
          this.statusDetail = '你已離開伺服器';
        }
        break;
      }
      case NetMessage.Pong: {
        const sentAt = decodeControl<{ t: number }>(bytes).payload.t;
        const sample = (performance.now() - sentAt) / 1000;
        // Smoothed, because one slow packet is not a slow connection and a ping
        // readout that jumps around is worse than no readout.
        this.rtt = this.rtt === 0 ? sample : this.rtt * 0.8 + sample * 0.2;
        break;
      }
      default:
        break;
    }
  }

  // -------------------------------------------------------------------------

  /**
   * Advance one tick: predict locally, then tell the server what was pressed.
   *
   * Called from the client's fixed-timestep loop, so this runs at exactly the
   * rate the server ticks at.
   */
  tick(local: PlayerState | undefined, input: InputCommand): void {
    if (this.status !== 'playing' || !this.socket) return;

    input.seq = this.seq++;
    if (local && local.alive) this.predictor.predict(local, input);

    this.send(encodeInputs(this.predictor.unacknowledged().map(toWire)));

    this.pingTimer -= TICK_DT;
    if (this.pingTimer <= 0) {
      this.pingTimer = NET.heartbeatInterval;
      this.send(encodeControl(NetMessage.Ping, { t: performance.now() }));
    }
  }

  /**
   * Fold in whatever the server has said since the last call.
   *
   * Separate from `tick` because it is driven by arrivals rather than by the
   * clock: a snapshot lands when it lands.
   */
  reconcile(local: PlayerState | undefined): void {
    const snap = this.snapshots.latest;
    if (!snap || !local) return;
    const mine = snap.players.find((p) => p.id === this.localId);
    if (!mine) return;

    // A dead player is not predicting anything, and replaying movement through a
    // respawn would fight the server over where they came back.
    if (!mine.alive || !local.alive) {
      this.predictor.reset();
      return;
    }

    this.predictor.reconcile(local, mine, snap.ackedInput);
  }

  /** Remote players, interpolated for this frame. */
  remotePlayers(dt: number): ReturnType<SnapshotBuffer['sample']> {
    this.sinceSnapshot += dt;
    return this.snapshots.sample(this.sinceSnapshot);
  }

  drainEvents(): SimEvent[] {
    const out = this.events.slice();
    this.events.length = 0;
    return out;
  }

  requestRespawn(): void {
    this.send(encodeControl(NetMessage.Respawn, {}));
  }

  say(text: string): void {
    this.send(encodeControl(NetMessage.Chat, { from: this.localId, text }));
  }

  stats(): PredictionStats & { ping: number; snapshots: number } {
    return {
      ...this.predictor.stats(),
      ping: Math.round(this.rtt * 1000),
      snapshots: this.snapshots.size,
    };
  }

  private send(bytes: Uint8Array): void {
    if (this.socket?.readyState === WebSocket.OPEN) this.socket.send(bytes);
  }

  dispose(): void {
    this.socket?.close();
    this.socket = null;
    this.snapshots.clear();
    this.predictor.reset();
  }
}

function toWire(c: InputCommand): WireInput {
  return {
    seq: c.seq,
    tick: c.tick,
    dt: c.dt,
    moveForward: c.moveForward,
    moveRight: c.moveRight,
    yaw: c.yaw,
    pitch: c.pitch,
    buttons: c.buttons,
    weaponSlot: 0,
  };
}

File: dotnet/src/OperationVanguard.Core/Networking/NetClient.cs
using System.Diagnostics;
using System.Text.Json;

namespace OperationVanguard.Core;

public enum NetStatus
{
    Connecting = 0,
    Playing = 1,
    Rejected = 2,
    Disconnected = 3,
}

/// <summary>
/// Minimal ordered-message transport contract. A WebSocket adapter can implement
/// this in the presentation layer; Core remains independent of any socket stack.
/// </summary>
public interface INetClientTransport
{
    bool IsOpen { get; }

    event Action? Opened;
    event Action<byte[]>? MessageReceived;
    event Action? Closed;
    event Action? Error;

    void Open(string url);
    void Send(byte[] bytes);
    void Close();
}

public interface INetClock
{
    double NowMilliseconds { get; }
}

/// <summary>A performance.now-like monotonic clock for production transports.</summary>
public sealed class MonotonicNetClock : INetClock
{
    private readonly long _origin = Stopwatch.GetTimestamp();

    public double NowMilliseconds => Stopwatch.GetElapsedTime(_origin).TotalMilliseconds;
}

public sealed class NetClientOptions
{
    public string Url { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public Loadout Loadout { get; set; } = new();
    public ICollisionWorld Collision { get; set; } = null!;
}

public readonly record struct NetClientStats(
    int Pending,
    int Mispredictions,
    double LastCorrection,
    int Ping,
    int Snapshots);

/// <summary>
/// Client-side connection state: immediate local movement prediction, remote
/// interpolation, reconciliation, control messages, and smoothed round-trip time.
/// </summary>
public sealed class NetClient : IDisposable
{
    private sealed class TimePayload
    {
        public double T { get; set; }
    }

    private readonly NetClientOptions _options;
    private Predictor _predictor;
    private readonly INetClock _clock;
    private INetClientTransport? _transport;
    private int _sequence = 1;
    private double _rtt;
    private double _pingTimer;
    private double _sinceSnapshot;
    private bool _disposed;

    public NetClient(
        NetClientOptions options,
        INetClientTransport transport,
        INetClock? clock = null)
    {
        _options = options ?? throw new ArgumentNullException(nameof(options));
        ArgumentNullException.ThrowIfNull(options.Collision);
        _transport = transport ?? throw new ArgumentNullException(nameof(transport));
        _clock = clock ?? new MonotonicNetClock();
        _predictor = new Predictor(options.Collision);

        transport.Opened += HandleOpened;
        transport.MessageReceived += HandleMessage;
        transport.Closed += HandleClosed;
        transport.Error += HandleError;
        transport.Open(options.Url);
    }

    public NetStatus Status { get; private set; } = NetStatus.Connecting;

    /// <summary>Why the connection ended, when it did.</summary>
    public string StatusDetail { get; private set; } = string.Empty;

    public WelcomePayload? Welcome { get; private set; }

    public int LocalId { get; private set; }

    public SnapshotBuffer Snapshots { get; } = new();

    public List<ChatPayload> Chat { get; } = [];

    /// <summary>Events received from the server, drained once per presentation frame.</summary>
    public List<SimEvent> Events { get; } = [];

    /// <summary>
    /// Replace the collision world used for local prediction without reopening
    /// the transport. Online presentation calls this once after Welcome when the
    /// server-selected map differs from the provisional menu selection.
    /// </summary>
    public void RebindCollision(ICollisionWorld collision)
    {
        ArgumentNullException.ThrowIfNull(collision);
        _options.Collision = collision;
        _predictor = new Predictor(collision);
    }

    /// <summary>
    /// Predict one fixed tick and send the complete unacknowledged input window.
    /// The protocol encoder retains only its newest sixteen entries on the wire.
    /// </summary>
    public void Tick(PlayerState? local, InputCommand input)
    {
        ArgumentNullException.ThrowIfNull(input);
        if (Status != NetStatus.Playing || _transport is null)
        {
            return;
        }

        input.Seq = _sequence++;
        if (local is { Alive: true })
        {
            _predictor.Predict(local, input);
        }

        var pending = _predictor.Unacknowledged();
        var wire = new List<WireInput>(pending.Count);
        foreach (var command in pending)
        {
            wire.Add(ToWire(command));
        }

        Send(NetProtocol.EncodeInputs(wire));

        _pingTimer -= GameConstants.TickDt;
        if (_pingTimer <= 0d)
        {
            _pingTimer = GameConstants.Network.HeartbeatInterval;
            Send(NetProtocol.EncodeControl(NetMessage.Ping, new TimePayload
            {
                T = _clock.NowMilliseconds,
            }));
        }
    }

    /// <summary>Fold the newest authoritative local-player frame into prediction.</summary>
    public void Reconcile(PlayerState? local)
    {
        var snapshot = Snapshots.Latest;
        if (snapshot is null || local is null)
        {
            return;
        }

        var authoritative = snapshot.Players.Find(player => player.Id == LocalId);
        if (authoritative is null)
        {
            return;
        }

        if (!authoritative.Alive || !local.Alive)
        {
            _predictor.Reset();
            return;
        }

        _predictor.Reconcile(local, authoritative, snapshot.AckedInput);
    }

    /// <summary>All players at this frame's delayed interpolation time.</summary>
    public IReadOnlyList<PlayerSnapshot> RemotePlayers(double deltaTime)
    {
        _sinceSnapshot += deltaTime;
        return Snapshots.Sample(_sinceSnapshot);
    }

    public List<SimEvent> DrainEvents()
    {
        var output = Events.ToList();
        Events.Clear();
        return output;
    }

    public void RequestRespawn() =>
        Send(NetProtocol.EncodeControl(NetMessage.Respawn, new Dictionary<string, object?>()));

    public void Say(string text) =>
        Send(NetProtocol.EncodeControl(NetMessage.Chat, new ChatPayload
        {
            From = LocalId,
            Text = text,
        }));

    public NetClientStats Stats()
    {
        var prediction = _predictor.Stats();
        return new NetClientStats(
            prediction.Pending,
            prediction.Mispredictions,
            prediction.LastCorrection,
            JsRound(_rtt * 1000d),
            Snapshots.Size);
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        var transport = _transport;
        if (transport is not null)
        {
            try
            {
                transport.Close();
            }
            finally
            {
                transport.Opened -= HandleOpened;
                transport.MessageReceived -= HandleMessage;
                transport.Closed -= HandleClosed;
                transport.Error -= HandleError;
                _transport = null;
            }
        }

        if (Status != NetStatus.Rejected)
        {
            Status = NetStatus.Disconnected;
            if (string.IsNullOrEmpty(StatusDetail))
            {
                StatusDetail = "連線中斷";
            }
        }

        Snapshots.Clear();
        _predictor.Reset();
    }

    private void HandleOpened()
    {
        Send(NetProtocol.EncodeControl(NetMessage.Hello, new HelloPayload
        {
            ProtocolVersion = GameConstants.Network.ProtocolVersion,
            Name = _options.Name,
            Loadout = _options.Loadout,
        }));
    }

    private void HandleMessage(byte[] bytes)
    {
        switch (NetProtocol.PeekType(bytes))
        {
            case NetMessage.Welcome:
                Welcome = NetProtocol.DecodeControl<WelcomePayload>(bytes).Payload;
                LocalId = Welcome.YourId;
                Status = NetStatus.Playing;
                break;

            case NetMessage.Reject:
                Status = NetStatus.Rejected;
                StatusDetail = NetProtocol.DecodeControl<RejectPayload>(bytes).Payload.Reason;
                break;

            case NetMessage.Snapshot:
                Snapshots.Push(NetProtocol.DecodeSnapshot(bytes));
                _sinceSnapshot = 0d;
                break;

            case NetMessage.Events:
                foreach (var simulationEvent in NetworkEventDecoder.Decode(bytes))
                {
                    Events.Add(simulationEvent);
                }
                break;

            case NetMessage.Chat:
                Chat.Add(NetProtocol.DecodeControl<ChatPayload>(bytes).Payload);
                if (Chat.Count > 32)
                {
                    Chat.RemoveAt(0);
                }
                break;

            case NetMessage.Bye:
                {
                    var bye = NetProtocol.DecodeControl<ByePayload>(bytes).Payload;
                    if (bye.Id == LocalId)
                    {
                        Status = NetStatus.Disconnected;
                        StatusDetail = "你已離開伺服器";
                    }
                    break;
                }

            case NetMessage.Pong:
                {
                    var payload = NetProtocol.DecodeControl<JsonElement>(bytes).Payload;
                    var sentAt = payload.GetProperty("t").GetDouble();
                    var sample = (_clock.NowMilliseconds - sentAt) / 1000d;
                    _rtt = _rtt == 0d ? sample : _rtt * 0.8d + sample * 0.2d;
                    break;
                }

            default:
                break;
        }
    }

    private void HandleClosed()
    {
        if (Status == NetStatus.Rejected)
        {
            return;
        }

        Status = NetStatus.Disconnected;
        if (string.IsNullOrEmpty(StatusDetail))
        {
            StatusDetail = "連線中斷";
        }
    }

    private void HandleError()
    {
        if (Status != NetStatus.Connecting)
        {
            return;
        }

        Status = NetStatus.Rejected;
        StatusDetail = "無法連線到伺服器";
    }

    private void Send(byte[] bytes)
    {
        if (_transport is { IsOpen: true } transport)
        {
            transport.Send(bytes);
        }
    }

    private static WireInput ToWire(InputCommand command) => new()
    {
        Seq = unchecked((uint)command.Seq),
        Tick = unchecked((uint)command.Tick),
        Dt = command.Dt,
        MoveForward = command.MoveForward,
        MoveRight = command.MoveRight,
        Yaw = command.Yaw,
        Pitch = command.Pitch,
        Buttons = unchecked((uint)command.Buttons),
        WeaponSlot = 0,
    };

    private static int JsRound(double value) => checked((int)Math.Floor(value + 0.5d));
}
